#!/usr/bin/env Rscript

suppressPackageStartupMessages(library(optparse))
arguments = parse_args(OptionParser(usage = "Rscript %prog [options]", option_list = list(make_option(c("-j", "--join_file"), dest = "join_file", help = "join.txt generated by home-made python script"), 
                                                                                      make_option(c("-i", "--input"), dest = "input", help = "generated by count_reads.py"),
                                                                                      make_option(c("-o", "--output"), dest = "output", help = "result"),
                                                                                      make_option(c("-e", "--fpkm_file"), dest = "fpkm_file", help = "FPKM values determined using StringTie (generated by parameter -A). "),
                                                                                      make_option(c("-c", "--mincov"), dest = "mincov", help = "skip all introns with coverage(%) lower than the given minimum value (default: 50)", default = 50)
                                                                                      )), 
                       positional_arguments = FALSE)

suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(reshape2))
suppressPackageStartupMessages(library(magrittr))

fpkm = read.table(arguments$fpkm_file, sep = '\t', header = T) %>% filter(FPKM >= 1)
filtered_gene_list = fpkm$Gene.Name

tb_intron_join = read.table(arguments$join_file, sep = '\t')
colnames(tb_intron_join) = c("Gene", "feature", "rank", "position", "length", "exon_rank_left", "exon_rank_right")
a = read.table(arguments$input, sep = '\t', header = T)

exon_a = a %>% filter(feature %in% c("exon")) %>% dplyr::select(-(feature)) %>% dplyr::rename("exon_rank" = rank)

res = a %>% filter(feature %in% c("intron")) %>% left_join(tb_intron_join, by = c("Gene", "feature", "rank", "position", "length")) %>% 
  
  dplyr::rename(exon_rank = exon_rank_left) %>% left_join(exon_a, by = c("Gene", "exon_rank"), suffix = c("", ".exon_left")) %>% select(-(exon_rank)) %>% 
  
  dplyr::rename(exon_rank = exon_rank_right) %>% left_join(exon_a, by = c("Gene", "exon_rank"), suffix = c("", ".exon_right")) %>% select(-(exon_rank)) %>% 
  
  mutate(gene_symbol = sapply(Gene, function(x) {strsplit(x, "/") %>% unlist %>% .[2]} )) %>% mutate(expr_state = case_when(
    gene_symbol %in% filtered_gene_list ~ "ok", 
    !(gene_symbol %in% filtered_gene_list) ~ "low expression"
  )) %>% 
  select(-(gene_symbol))

# debug: add rowwise()
res %>% filter(coverage... >= arguments$mincov) %>% rowwise() %>% mutate(iri = c(read_counts_norm.exon_left, read_counts_norm.exon_right) %>% na.omit %>% mean %>% {read_counts_norm/(.)}) %>% write.table(file = arguments$output, sep = '\t', quote = F, row.names = F)





